typescript/packages/lookslike-high-level/src/components/iframe-view.ts

```diff
@@ -0,0 +1,97 @@
+import { LitElement, html } from "lit-element";
+import { customElement, property } from "lit/decorators.js";
+import { Cell } from "@commontools/common-runner";
+import { Ref, createRef, ref } from "lit/directives/ref.js";
+
+@customElement("common-iframe")
+export class CommonIframe extends LitElement {
+  @property({ type: String }) src = "";
+  // HACK: The UI framework already translates the top level cell into updated
+  // properties, but we want to only have to deal with one type of listening, so
+  // we'll add a an extra level of indirection with the "context" property.
+  @property({ type: Object }) context?: Cell<any> | any;
+
+  private iframeRef: Ref<HTMLIFrameElement> = createRef();
+
+  private subscriptions: Map<string, () => void> = new Map();
+
+  private handleMessage = (event: MessageEvent) => {
+    console.log("Received message", event);
+    if (event.source === this.iframeRef.value?.contentWindow) {
+      const { type, key, data } = event.data;
+      if (typeof key !== "string") {
+        console.error("Invalid key type. Expected string.");
+        return;
+      }
+      if (type === "read" && this.context) {
+        const value = this.context?.getAsProxy
+          ? this.context?.getAsProxy([key])
+          : this.context?.[key];
+        // TODO: This might cause infinite loops, since the data can be a graph.
+        const copy = JSON.parse(JSON.stringify(value ?? {}));
+        console.log("readResponse", key, value);
+        this.iframeRef.value?.contentWindow?.postMessage(
+          { type: "readResponse", key, data: copy },
+          "*"
+        );
+      } else if (type === "write" && this.context) {
+        this.context.getAsProxy()[key] = data;
+      } else if (type === "subscribe" && this.context) {
+        if (!this.subscriptions) {
+          this.subscriptions = new Map();
+        }
+        console.log("subscribing", key, this.context);
+        // TODO(ben): should be subscribing to just the cell, not the whole context
+        // but doing .getAsProxy([key]) or [key] both seem to be lacking .sink()?
+        const unsub = this.context.sink(() => {
+          this.notifySubscribers(key, this.context.get()[key]);
+        });
+        this.subscriptions.set(key, unsub);
```

Comment:
@seefeldb not sure what I'm missing here, this code works because it's triggering for _any_ change but I wanted to scope to just a single key... Is it because the value is a primitive (number)? Am I just holding it wrong?

Reply:
/me really needs to write that README :)

.getAsProxy() turns it into a proxy that just represents the data, but also traverses cell references, since our data can be a graph.

`this.context.key(key).sink()` would do what you want, but it won't subscribe to changes in referenced cells.

what we can do is call `.getAsProxy([key], log)` where log is a `ReactivityLog`. And then do the deep copy. Then `log` contains all the cells actually read.

But at this point, we can also use the actual scheduler to push changes. Let me do this quickly, might need a few small tweaks.